share/vss.go

```diff
@@ -0,0 +1,493 @@
+	if err := d.verifyComplaint(c); err != nil {
+		return nil, err
+	}
+	// index is guaranteed to be found because verifyComplaint does the check
+	i, _ := findIndex(d.verifiers, c.Public)
+	return &DealerResponse{
+		Deal:      d.deals[i],
+		Complaint: c,
+	}, nil
+}
+
+// ReceiveApproval looks if the approval is legitimate and stores it. If the
+// approval is not legitimate or has already been received, it returns an error.
+// To knoew whether participants approved the sharing, call `d.EnoughApproval`
+// and if true, `d.DealCertified`.
+func (d *Dealer) ReceiveApproval(a *Approval) error {
+	return d.verifyApproval(a)
+}
+
+// Verifier receives a Deal from a Dealer, can reply by a Complaint, and can
+// collaborate with other Verifiers to reconstruct a secret.
+type Verifier struct {
+	suite abstract.Suite
+
+	long abstract.Scalar
+	pub  abstract.Point
+
+	dealer abstract.Point
+
+	index     int
+	verifiers []abstract.Point
+
+	h abstract.Point
+
+	commitPoly *PubPoly
+
+	fi *PriShare
+	gi *PriShare
+
+	*aggregator
+
+	dealerBad bool
+}
+
+// NewVerifier returns a Verifier out of:
+// - its longterm secret key
+// - the longterm dealer public key
+// - the list of public key of verifiers. The list MUST include the public key
+// of this Verifier also.
+// The security parameter t of the secret sharing scheme is automatically set to
+// a default safe value. If a different t value is required, it is possible to set
+// it with `verifier.SetT()`.
+func NewVerifier(suite abstract.Suite, longterm abstract.Scalar, dealerKey abstract.Point,
+	verifiers []abstract.Point) (*Verifier, error) {
+	v := &Verifier{
+		suite:     suite,
+		long:      longterm,
+		dealer:    dealerKey,
+		verifiers: verifiers,
+	}
+	v.pub = v.suite.Point().Mul(nil, v.long)
+	var ok bool
+	if v.index, ok = findIndex(verifiers, v.pub); !ok {
+		return nil, errors.New("verifier: not in the list of verifiers")
+	}
+	v.h = deriveH(suite, verifiers)
+	return v, nil
+}
+
+// ReceiveDeal takes a Deal, tries to decrypt it and see if it is correct or
+// not. In case the Deal is correct,ie. the verifier can verify the shares
+// against the public coefficients, an Approval is returned and must be
+// broadcasted to every participants including the Dealer. In case the Deal
+// is not correct because of the public polynomial checking, or the indexes are
+// not the same, or the session identifier is wrong, it returns a Complaint
+// which must be broadcasted to every participants including the Dealer.
+// If the complaint is deemed invalid because of a wrong index (not the same as
+// the verifier) or because the verifier already received a Deal, it does not
+// return a complaint.
+// ReceiveDeal returns an error in any case to let the user know what was the
+// error.
+// XXX API question: return value.For the moment, the default is to let the user
+// know everything that is wrong through the error.
+func (v *Verifier) ReceiveDeal(d *Deal) (*Approval, *Complaint, error) {
```

Comment:
Would it make sense to replace the "Receive" prefixes everywhere with "Process" or so? To me "Receive" always implicates some involvement of network communication but at this point the receipt of the message/deal has already happened at another place.